x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java

```diff
@@ -214,7 +213,176 @@ private Aggs addPipelineAggs(QueryContainer qContainer, QueryTranslation qt, Fil
+
+                tail = lastAgg;
+            }
+
+            GroupByKey groupFor(Expression exp) {
+                Integer hash = null;
+                if (Functions.isAggregate(exp)) {
+                    AggregateFunction f = (AggregateFunction) exp;
+                    // if there's at least one agg in the tree
+                    if (groupMap.isEmpty() == false) {
+                        GroupByKey matchingGroup = null;
+                        // group found - finding the dedicated agg
+                        // TODO: when dealing with expressions inside Aggregation, make sure to extract the field
+                        hash = Integer.valueOf(f.field().hashCode());
+                        matchingGroup = groupMap.get(hash);
+                        // return matching group or the tail (last group)
+                        return matchingGroup != null ? matchingGroup : tail;
+                    } else {
+                        return null;
+                    }
+                }
+
+                hash = Integer.valueOf(exp.hashCode());
+                return groupMap.get(hash);
+            }
+
+            @Override
+            public String toString() {
+                return groupMap.toString();
+            }
+        }
+
+        /**
+         * Creates the list of GroupBy keys
+         */
+        static GroupingContext groupBy(List<? extends Expression> groupings) {
+            if (groupings.isEmpty() == true) {
+                return null;
+            }
+
+            Map<Integer, GroupByKey> aggMap = new LinkedHashMap<>();
+
+            for (Expression exp : groupings) {
+                GroupByKey key = null;
+
+                Integer hash = Integer.valueOf(exp.hashCode());
+                String aggId = Expressions.id(exp);
+
+                // change analyzed to non non-analyzed attributes
+                if (exp instanceof FieldAttribute) {
+                    FieldAttribute field = (FieldAttribute) exp;
+                    field = field.exactAttribute();
+                    key = new GroupByValue(aggId, field.name());
+                }
+
+                // handle functions
+                else if (exp instanceof Function) {
+                    // dates are handled differently because of date histograms
+                    if (exp instanceof DateTimeHistogramFunction) {
+                        DateTimeHistogramFunction dthf = (DateTimeHistogramFunction) exp;
+
+                        Expression field = dthf.field();
+                        if (field instanceof FieldAttribute) {
+                            if (dthf.calendarInterval() != null) {
+                                key = new GroupByDateHistogram(aggId, QueryTranslator.nameOf(exp), dthf.calendarInterval(), dthf.zoneId());
+                            } else {
+                                key = new GroupByDateHistogram(aggId, QueryTranslator.nameOf(exp), dthf.fixedInterval(), dthf.zoneId());
+                            }
```

Comment:
No, because each method returns a different type - the first a `String`, the second a `long` so the common type would be `Object` at which point is unclear what constructor to call...

Reply:
You're right. Sorry, I missed that.